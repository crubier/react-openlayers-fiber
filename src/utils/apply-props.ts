import { startsWith } from "lodash/fp";
import OlObject from "ol/Object";

/**
 * Determines if an object as a setter for a given key
 *
 * @param object
 * @param key
 * @returns (boolean) `true` if the object has a setter for the given key.
 * else return `false`
 */
export const hasSetter = (key: string, object: object): boolean =>
  Object.getOwnPropertyDescriptor(object, key)?.set != null;

/**
 *
 * This code checks, for every given props,
 * if the ol entity has a setter for the prop.
 * If it has one, it sets the value to the ol object,
 * but it only sets it if it changed from the previous props.
 *
 * @param olObject The ol object to update
 * @param newProps The newProps potentially containing new changes
 */
export const applyProps = (
  olObject: OlObject,
  oldProps: object = {},
  newProps: object
): void => {
  const olObjectKeys = olObject.getKeys();
  Object.entries(newProps)
    // .filter(([key]) => includes(key, olObjectKeys))
    .forEach(([key, value]) => {
      if (oldProps[key] !== newProps[key]) {
        // For special cases (for example ol objects that have an option called "key"), we can add a "_" before.
        const olKey = startsWith("_", key) ? key.substring(1) : key;
        olObject.set(olKey, value);
      }
    });
};
