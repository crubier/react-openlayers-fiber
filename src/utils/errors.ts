const header = (number) => `[react-openlayers-fiber] Error${number} - `;

export const error001 = new Error(header("001") + 'Unsupported "attach" type.');
export const error002 = (containerType, childType) =>
  new Error(
    header("002") +
      `Couldn't add this child to this container. You can specify how to attach this type of child ("${childType}") to this type of container ("${containerType}") using the "attach" props.`
  );
export const error003 = (containerType, childType) =>
  new Error(
    header("003") +
      `Couldn't remove this child from this container. You can specify how to detach this type of child ("${childType}") from this type of container ("${containerType}") using the "attach" props.`
  );
export const error004 = (name) =>
  new Error(header("004") + `${name} is not exported by ol.`);
export const error005 = (constructorName, objectName) =>
  new Error(
    header("005") + `${constructorName} is not a constructor for ${objectName}`
  );
