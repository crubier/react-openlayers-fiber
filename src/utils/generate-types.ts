///////////////////////////////////////////////////////////////////////////////
// To run this you need to comment
//
// in "ol/layer.js":
// export { default as Heatmap } from './layer/Heatmap.js';
// export { default as WebGLPoints } from './layer/WebGLPoints.js';
//
// in "ol/source.js":
// export { default as Raster } from './source/Raster.js';

import {
  map,
  join,
  split,
  head,
  upperCase,
  camelCase,
  filter,
  keys,
  sortBy,
  identity,
  upperFirst,
  pick,
} from "lodash/fp";

///////////////////////////////////////////////////////////////////////////////
// Imports of elements
///////////////////////////////////////////////////////////////////////////////
// ol: Special case because we don't want all ol stuff
import * as olTemp from "ol";
const ol = pick(["Map", "View", "Feature", "Collection"], olTemp);
// layer: Special case because some layer types cause problems during code generation...
import * as layerTemp from "ol/layer";
const layer = { ...layerTemp, WebGLPoints: {}, Heatmap: {} };
// control: all good
import * as control from "ol/control";
// interaction: all good
import * as interaction from "ol/interaction";
// source: Special case because some source types cause problems during code generation...
import * as sourceTemp from "ol/source";
const source = { ...sourceTemp, Raster: {} };
// geom: all good
import * as geom from "ol/geom";
// overlay: Special case because there is only one type, so it does not have the same shape a others
import Overlay from "ol/Overlay";
const overlay = { Overlay };
///////////////////////////////////////////////////////////////////////////////

import { writeFileSync } from "fs";

export const isFirstLetterCapitalized = (key: string) => {
  const firstLetter = head(split("", key));
  return upperCase(firstLetter) === firstLetter;
};

const filterCapitalized = filter(isFirstLetterCapitalized);

export const convertMappingExported = (prefix: string) => (
  key: string
): string => {
  return `  ${prefix}${upperFirst(camelCase(key))}: typeof ${prefix}.${key};`;
};

export const convertMapping = (prefix: string) => (key: string): string => {
  return `  ${prefix}${upperFirst(camelCase(key))}: ${prefix}.${key},`;
};

export const convertKindMapping = (prefix: string) => (key: string): string => {
  return `  ${prefix}${upperFirst(camelCase(key))}: "${prefix==="ol"?upperFirst(key):upperFirst(prefix)}",`;
};

export const convertMappingTypeOf = (prefix: string) => (
  key: string
): string => {
  return `  ${prefix}${upperFirst(
    camelCase(key)
  )}: typeof ${prefix}["${key}"];`;
};

export const convertAllValues = () => {
  return `// @ts-nocheck
// Generated Code, do not edit manually
import * as ol from "ol";
import * as layer from "ol/layer";
import * as control from "ol/control";
import * as interaction from "ol/interaction";
import * as source from "ol/source";
import * as geom from "ol/geom";
import Overlay from "ol/Overlay";
const overlay = {Overlay};

export type MappingExported = {
${join(
  "\n",
  sortBy(identity, [
    ...map(convertMappingExported("ol"), filterCapitalized(keys(ol))),
    ...map(convertMappingExported("layer"), filterCapitalized(keys(layer))),
    ...map(convertMappingExported("control"), filterCapitalized(keys(control))),
    ...map(
      convertMappingExported("interaction"),
      filterCapitalized(keys(interaction))
    ),
    ...map(convertMappingExported("source"), filterCapitalized(keys(source))),
    ...map(convertMappingExported("geom"), filterCapitalized(keys(geom))),
    ...map(convertMappingExported("overlay"), filterCapitalized(keys(overlay))),
  ])
)}
};

export const mapping = {
${join(
  "\n",
  sortBy(identity, [
    ...map(convertMapping("ol"), filterCapitalized(keys(ol))),
    ...map(convertMapping("layer"), filterCapitalized(keys(layer))),
    ...map(convertMapping("control"), filterCapitalized(keys(control))),
    ...map(convertMapping("interaction"), filterCapitalized(keys(interaction))),
    ...map(convertMapping("source"), filterCapitalized(keys(source))),
    ...map(convertMapping("geom"), filterCapitalized(keys(geom))),
    ...map(convertMapping("overlay"), filterCapitalized(keys(overlay))),
  ])
)}
};

export const kindMapping = {
${join(
  "\n",
  sortBy(identity, [
    ...map(convertKindMapping("ol"), filterCapitalized(keys(ol))),
    ...map(convertKindMapping("layer"), filterCapitalized(keys(layer))),
    ...map(convertKindMapping("control"), filterCapitalized(keys(control))),
    ...map(convertKindMapping("interaction"), filterCapitalized(keys(interaction))),
    ...map(convertKindMapping("source"), filterCapitalized(keys(source))),
    ...map(convertKindMapping("geom"), filterCapitalized(keys(geom))),
    ...map(convertKindMapping("overlay"), filterCapitalized(keys(overlay))),
  ])
)}
};

export type MappingTypeofExport = {
${join(
  "\n",
  sortBy(identity, [
    ...map(convertMappingTypeOf("ol"), filterCapitalized(keys(ol))),
    ...map(convertMappingTypeOf("layer"), filterCapitalized(keys(layer))),
    ...map(convertMappingTypeOf("control"), filterCapitalized(keys(control))),
    ...map(
      convertMappingTypeOf("interaction"),
      filterCapitalized(keys(interaction))
    ),
    ...map(convertMappingTypeOf("source"), filterCapitalized(keys(source))),
    ...map(convertMappingTypeOf("geom"), filterCapitalized(keys(geom))),
    ...map(convertMappingTypeOf("overlay"), filterCapitalized(keys(overlay))),
  ])
)}
};`;
};

// We don't want to test those functions
/* istanbul ignore next */
/**
 * This functions generates a file containing 2 mappings necessary in order to
 * have proper typing of react-openlayers-fiber elements.
 *
 * It needs to be updated on every release of ol.
 * Maybe we should put those types in a different package in order to avoid tying
 * react-openlayers-fiber releases with ol ones.
 */
export const generateTypes = () =>
  writeFileSync("./src/generated-mapping.ts", convertAllValues());

/* istanbul ignore next */
if (require.main === module) {
  generateTypes();
  console.log("mapping types generated!");
}
