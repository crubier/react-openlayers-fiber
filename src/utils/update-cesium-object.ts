/**
 * Determines if an object as a setter for a given key
 *
 * @param object
 * @param key
 * @returns (boolean) `true` if the object has a setter for the given key.
 * else return `false`
 */
export const hasSetter = (key: string, object: object): boolean =>
  Object.getOwnPropertyDescriptor(object, key)?.set != null;

export const forEachSetterOf = (
  iteratee: (key: string, value: any) => void,
  referenceObject: object,
  object: object
): void => {
  Object.entries(object)
    .filter(([key]) => hasSetter(key, referenceObject))
    .forEach(([key, value]) => iteratee(key, value));
};

/**
 *
 * This code checks, for every given props,
 * if the cesium entity has a setter for the prop.
 * If it has one, it sets the value to the cesium object,
 * but it only sets it if it changed from the previous props.
 *
 * @param cesiumObject The cesium object to update
 * @param props The props potentially containing new changes
 */
export const updateCesiumObject = (
  cesiumObject: object,
  oldProps: object = {},
  props: object
): void => {
  forEachSetterOf(
    (key, value) => {
      if (oldProps[key] !== value) {
        cesiumObject[key] = value;
      }
    },
    Object.getPrototypeOf(cesiumObject),
    props
  );
};
