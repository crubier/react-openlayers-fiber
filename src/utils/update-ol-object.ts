import { includes } from "lodash/fp";
import OlObject from "ol/Object";

/**
 * Determines if an object as a setter for a given key
 *
 * @param object
 * @param key
 * @returns (boolean) `true` if the object has a setter for the given key.
 * else return `false`
 */
export const hasSetter = (key: string, object: object): boolean =>
  Object.getOwnPropertyDescriptor(object, key)?.set != null;

/**
 *
 * This code checks, for every given props,
 * if the ol entity has a setter for the prop.
 * If it has one, it sets the value to the ol object,
 * but it only sets it if it changed from the previous props.
 *
 * @param olObject The ol object to update
 * @param props The props potentially containing new changes
 */
export const updateOlObject = (
  olObject: OlObject,
  oldProps: object = {},
  props: object
): void => {
  const olObjectKeys = olObject.getKeys();
  Object.entries(props)
    .filter(([key]) => includes(key, olObjectKeys))
    .forEach(([key, value]) => {
      if (oldProps[key] !== value) {
        olObject.set(key, value);
      }
    });
};
