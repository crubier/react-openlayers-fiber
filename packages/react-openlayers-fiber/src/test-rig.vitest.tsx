import { render,screen, userEvent } from "./test/utils";
import { Map } from "ol";
import React from "react";
import { describe, expect, it } from 'vitest'



// describe('Simple working test', () => {
//   it('the title is visible', () => {
//     render(<App />)
//     expect(screen.getByText(/Hello Vite \+ React!/i)).toBeInTheDocument()
//   })

//   it('should increment count on click', async() => {
//     render(<App />)
//     userEvent.click(screen.getByRole('button'))
//     expect(await screen.findByText(/count is: 1/i)).toBeInTheDocument()
//   })
// })



describe("Test setup", () => {
  it("can run basic test", () => {
    expect(1).toBe(1);
  });
  it("can run inline snapshot test", () => {
    expect(1).toMatchInlineSnapshot(`1`);
  });
  it("can run dom snapshot test", () => {
    const { container } = render(<div>ok</div>);
    expect(container).toMatchSnapshot();
  });
  it("can run react testing library test", () => {
    const { getByText } = render(<div>ok</div>);
    const heading = getByText(/ok/i);
    expect(heading).toMatchInlineSnapshot(`
<div>
  ok
</div>
`);
  });
  it("can run ol tests", () => {
    const map = new Map({});
    expect(map.getAllLayers().length).toBe(0);
  });
});
