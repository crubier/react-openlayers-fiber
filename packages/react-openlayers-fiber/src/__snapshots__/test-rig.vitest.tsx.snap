// Vitest Snapshot v1

exports[`Test setup > can run dom snapshot test 1`] = `
<div>
  <div>
    ok
  </div>
</div>
`;
